sdk/log/provider_test.go

```diff
@@ -308,3 +308,18 @@ func BenchmarkLoggerProviderLogger(b *testing.B) {
 	b.StopTimer()
 	loggers[0].Enabled(context.Background(), log.Record{})
 }
+
+func BenchmarkLoggerProviderLoggerSame(b *testing.B) {
+	p := NewLoggerProvider()
+
+	b.ResetTimer()
+	b.ReportAllocs()
+
+	var logger log.Logger
+	for i := 0; i < b.N; i++ {
+		logger = p.Logger("test", log.WithInstrumentationVersion("v1.2.3"))
```

Comment:
Similar to our [metric benchmarks](https://github.com/open-telemetry/opentelemetry-go/blob/e8c22e6e7180056cbb229a474c8ee98c7696ce07/sdk/metric/benchmark_test.go#L48-L108), this passed option should be allocated outside of the evaluation loop. Otherwise, we are measuring the inefficient misuse of the API.

Reply:
This is the whole point of the benchmark is to show that such usage introduces a heap allocation.

I think that it would be better if the user would not need to preallocate instrumentation version in order not introduce a heap allocation.

Reply:
Maybe I should add a comment like: "Showcase that not pre-allocating the options slice causes a heap allocation"?

Reply:
If that is the case, then this seems like a re-hash of the already concluded extensive conversation about how we handle configuration:

- https://github.com/open-telemetry/opentelemetry-go/blob/main/CONTRIBUTING.md#configuration
- https://github.com/open-telemetry/opentelemetry-go/pull/971
- https://github.com/open-telemetry/opentelemetry-go/issues/536

Additionally, since we have already discussed this specific detail for the metric signal[^1][^2] extensively, unless there are new developments, I would prefer to not have the same discussion again. Especially since there is are now 2 stable signals that have set our precedence and a documented policy.

[^1]: https://github.com/open-telemetry/opentelemetry-go/pull/3971#discussion_r1163272609
[^2]: topic of discussion in the April 13 2023 SIG meeting: https://docs.google.com/document/d/1E5e7Ld0NuU1iVvf-42tOBpu2VBBLYnh73GJuITGJTTU/edit#heading=h.9lojwomaurj2

Reply:
I think we made a bad decision to use options in measurements and we should retrospect our decisions.

I am fine using options only for bootstrapping the SDK. However, I find that using it on hot-paths makes it too easy for the callers to write inefficient code.

I respect your opinion, but I would like also like to hear what other @open-telemetry/go-maintainers think about it.

Reply:
I am fine using options for things bootstrapping the SDK (creating exporters, providers, processors).

However, I find that using options on code that may be on the hot-path makes it too easy for the callers to write inefficient code. Take notice that `log.Record` already diverges from these guidelines. Maybe, we should work on an alternate guideline for handling configuration for code that we asses that may be on the hot-path (e.g. when recording a value)?

I am closing this PR and creating a new issue. Thank you for your insight.